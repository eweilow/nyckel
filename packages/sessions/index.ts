export { SessionManager, createSessionManager } from "./manager";
