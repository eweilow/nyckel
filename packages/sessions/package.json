{
  "name": "@nyckel/sessions",
  "version": "0.0.1-alpha.11",
  "description": "The module of Nyckel that provides secure sessions",
  "main": "./dist/cjs",
  "module": "./dist/esm",
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/eweilow/nyckel.git"
  },
  "author": "eweilow",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/eweilow/nyckel/issues"
  },
  "homepage": "https://github.com/eweilow/nyckel",
  "dependencies": {
    "aes256": "^1.0.4",
    "redis": "^2.8.0",
    "redlock": "^4.0.0",
    "uuid": "^3.3.2"
  },
  "devDependencies": {
    "@types/node": "12.7.1",
    "@types/redis": "2.8.13",
    "@types/redlock": "3.0.4",
    "@types/uuid": "8.3.4"
  },
  "keywords": [
    "nyckel"
  ],
  "types": "./",
  "gitHead": "48ceeacb01d90f875aa79ad1c6cfc593184f579b"
}
