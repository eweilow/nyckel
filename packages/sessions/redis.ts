import { createClient, ClientOpts } from "redis";
import { EventEmitter } from "events";

export function createRedisClient(
  endpoint: string,
  handleError: (err: Error) => void,
  options?: ClientOpts
) {
  const client = createClient(endpoint, options);

  const healthyEventEmitter = new EventEmitter();

  let unhealthyTimeout: NodeJS.Timer | null = null;
  function markUnhealthy() {
    if (unhealthyTimeout != null) {
      return;
    }
    unhealthyTimeout = setTimeout(() => {
      healthy = false;
      healthyEventEmitter.emit("unhealthy");
      unhealthyTimeout = null;
    }, 5000);
  }

  let healthy: boolean = true;
  function markHealthy() {
    if (unhealthyTimeout != null) {
      clearTimeout(unhealthyTimeout);
      unhealthyTimeout = null;
    }

    if (!healthy) {
      healthy = true;
      healthyEventEmitter.emit("healthy");
    }
  }

  client.on("error", err => {
    handleError(err);
  });

  client.on("end", () => {
    markUnhealthy();
  });

  client.on("reconnecting", () => {
    markUnhealthy();
  });

  client.on("ready", () => {
    markHealthy();
  });

  markUnhealthy();

  return {
    client,
    get healthy() {
      return healthy;
    },
    healthyEventEmitter
  };
}
