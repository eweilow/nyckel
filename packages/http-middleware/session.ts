import { SessionManager } from "@nyckel/sessions";

export type SessionInRequest = {
  delete(): Promise<void>;
  get<T = any>(): Promise<T>;
  set<T = any>(data: T): Promise<void>;
};

export function enhanceRequestWithSession<T>(
  request: T,
  cookieName: string,
  cookie: string,
  sessions: SessionManager,
  setCookie: (name: string, value: string) => void,
  clearCookie: (name: string) => void
): T & { session: SessionInRequest } {
  let enhancedRequest = request as T & { session: SessionInRequest };
  enhancedRequest.session = {
    async set(data) {
      let id = cookie;
      if (!sessions.isValidId(id)) {
        id = sessions.generateId();
      }
      setCookie(cookieName, id);
      await sessions.set(id, data);
    },
    async get() {
      const id = cookie;
      if (sessions.isValidId(id)) {
        const value = await sessions.get(id);
        if (value != null) {
          return value;
        }
      }
      return {};
    },
    async delete() {
      const id = cookie;
      if (sessions.isValidId(id)) {
        await sessions.delete(id);
        clearCookie(cookieName);
      }
    }
  };

  return enhancedRequest;
}
