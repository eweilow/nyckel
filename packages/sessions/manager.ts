import { createRedisClient } from "./redis";
import { ClientOpts } from "redis";
import { createKeyHash } from "./crypto/key";
import { encryptData, decryptData } from "./crypto/value";

import uuid from "uuid/v4";

import Redlock from "redlock";
import { wrapInPromise } from "./utils/wrapInPromise";

export type SessionManager = ReturnType<typeof createSessionManager>;

export function createSessionManager(
  redisEndpoint: string,
  handleError: (err: Error) => void,
  options: {
    secret: string;
    salt: string;
    ttl: number;
    redisOptions?: ClientOpts;
  }
) {
  const { client } = createRedisClient(
    redisEndpoint,
    handleError,
    options.redisOptions
  );

  const redlock = new Redlock([client], {
    retryCount: 0
  });

  return {
    get client() {
      return client;
    },
    isValidId(id?: string) {
      if (id == null) {
        return false;
      }
      if (typeof id !== "string") {
        return false;
      }
      if (id.length > 36) {
        return false;
      }
      return /^[a-z0-9\-]{36}$/i.test(id);
    },
    generateId() {
      return uuid();
    },
    async lock(
      id: string,
      ttl: number
    ): Promise<
      { acquired: true; release: () => Promise<void> } | { acquired: false }
    > {
      const key = createKeyHash(id, options.salt);
      try {
        const lock = await redlock.lock("locks:" + key, ttl);
        return {
          acquired: true,
          async release() {
            await lock.unlock();
          }
        };
      } catch (err) {
        return {
          acquired: false
        };
      }
    },
    async set<T = any>(id: string, data: any) {
      const key = createKeyHash(id, options.salt);

      const encrypted = encryptData<T>(id, options.salt, options.secret, data);
      await wrapInPromise(cb => client.setex(key, options.ttl, encrypted, cb));
    },
    async get<T = any>(id: string) {
      const key = createKeyHash(id, options.salt);

      const encrypted = await wrapInPromise<string>(cb => client.get(key, cb));
      if (encrypted == null) {
        return null;
      }

      await wrapInPromise(cb => client.expire(key, options.ttl, cb));

      return decryptData<T>(id, options.salt, options.secret, encrypted);
    },
    async delete(id: string) {
      const key = createKeyHash(id, options.salt);

      await wrapInPromise(cb => client.del(key, cb));
    },
    async has(id: string) {
      const key = createKeyHash(id, options.salt);

      return wrapInPromise(cb => client.exists(key, cb));
    }
  };
}
