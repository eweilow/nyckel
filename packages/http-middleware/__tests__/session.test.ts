import { enhanceRequestWithSession } from "../session";
import { SessionManager } from "@nyckel/sessions";

describe("enhanceRequestWithSession", () => {
  it("should enhance a request correctly", () => {
    let req = {};

    const setCookie = jest.fn();
    const clearCookie = jest.fn();

    let enhancedReq = enhanceRequestWithSession(
      req,
      "cookieName",
      "cookie",
      {} as any,
      setCookie,
      clearCookie
    );

    expect(enhancedReq.session).toBeDefined();
    expect(typeof enhancedReq.session.set).toBe("function");
    expect(typeof enhancedReq.session.get).toBe("function");
    expect(typeof enhancedReq.session.delete).toBe("function");
    expect(enhancedReq.session).toMatchSnapshot();

    expect(setCookie).not.toHaveBeenCalled();
    expect(clearCookie).not.toHaveBeenCalled();
  });

  const cookieName = "cookieName";

  describe(".set", () => {
    it("should generate an id if the provided cookie is invalid", async () => {
      const sessions: SessionManager = {
        delete: jest.fn(),
        get: jest.fn(),
        has: jest.fn(),
        lock: jest.fn(),
        set: jest.fn(),
        isValidId: jest.fn(id => id === "valid"),
        generateId: jest.fn(() => "generated")
      };

      const setCookie = jest.fn();
      const clearCookie = jest.fn();

      let enhancedReq = enhanceRequestWithSession(
        {},
        cookieName,
        "cookie",
        sessions,
        setCookie,
        clearCookie
      );

      let data = { test: "" };
      await enhancedReq.session.set(data);
      expect(setCookie).toHaveBeenCalledWith(cookieName, "generated");
      expect(clearCookie).not.toHaveBeenCalled();
      expect(sessions.generateId).toHaveBeenCalled();
      expect(sessions.set).toHaveBeenCalledWith("generated", data);
    });

    it("should not generate an id if the provided cookie is valid", async () => {
      const sessions: SessionManager = {
        delete: jest.fn(),
        get: jest.fn(),
        has: jest.fn(),
        lock: jest.fn(),
        set: jest.fn(),
        isValidId: jest.fn(id => id === "valid"),
        generateId: jest.fn(() => "generated")
      };

      const setCookie = jest.fn();
      const clearCookie = jest.fn();

      let enhancedReq = enhanceRequestWithSession(
        {},
        cookieName,
        "valid",
        sessions,
        setCookie,
        clearCookie
      );

      let data = { test: "" };
      await enhancedReq.session.set(data);
      expect(setCookie).toHaveBeenCalledWith(cookieName, "valid");
      expect(clearCookie).not.toHaveBeenCalled();
      expect(sessions.generateId).not.toHaveBeenCalled();
      expect(sessions.set).toHaveBeenCalledWith("valid", data);
    });
  });

  describe(".get", () => {
    it("should return empty object if the provided cookie is invalid", async () => {
      const sessions: SessionManager = {
        delete: jest.fn(),
        get: jest.fn(),
        has: jest.fn(),
        lock: jest.fn(),
        set: jest.fn(),
        isValidId: jest.fn(id => id === "valid"),
        generateId: jest.fn()
      };

      const setCookie = jest.fn();
      const clearCookie = jest.fn();

      let enhancedReq = enhanceRequestWithSession(
        {},
        cookieName,
        "invalid",
        sessions,
        setCookie,
        clearCookie
      );

      const data = await enhancedReq.session.get();
      expect(data).toEqual({});

      expect(sessions.get).not.toHaveBeenCalled();

      type T = string | null;
      expect(setCookie).not.toHaveBeenCalled();
      expect(clearCookie).not.toHaveBeenCalled();
    });

    it("should return empty object if the session state is invalid", async () => {
      const sessions: SessionManager = {
        delete: jest.fn(),
        get: jest.fn((async () => null) as any),
        has: jest.fn(),
        lock: jest.fn(),
        set: jest.fn(),
        isValidId: jest.fn(id => id === "valid"),
        generateId: jest.fn()
      };

      const setCookie = jest.fn();
      const clearCookie = jest.fn();

      let enhancedReq = enhanceRequestWithSession(
        {},
        cookieName,
        "valid",
        sessions,
        setCookie,
        clearCookie
      );

      const data = await enhancedReq.session.get();
      expect(data).toEqual({});

      expect(sessions.get).toHaveBeenCalledWith("valid");

      expect(setCookie).not.toHaveBeenCalled();
      expect(clearCookie).not.toHaveBeenCalled();
    });

    it("should return the session object if it's truthy", async () => {
      const sessions: SessionManager = {
        delete: jest.fn(),
        get: jest.fn(async id => ({ id, data: "yes" } as any)),
        has: jest.fn(),
        lock: jest.fn(),
        set: jest.fn(),
        isValidId: jest.fn(id => id === "valid"),
        generateId: jest.fn()
      };

      const setCookie = jest.fn();
      const clearCookie = jest.fn();

      let enhancedReq = enhanceRequestWithSession(
        {},
        cookieName,
        "valid",
        sessions,
        setCookie,
        clearCookie
      );

      const data = await enhancedReq.session.get();
      expect(data).toEqual({ id: "valid", data: "yes" });

      expect(sessions.get).toHaveBeenCalledWith("valid");

      expect(setCookie).not.toHaveBeenCalled();
      expect(clearCookie).not.toHaveBeenCalled();
    });
  });

  describe(".delete", () => {
    it("should do nothing the provided cookie is invalid", async () => {
      const sessions: SessionManager = {
        delete: jest.fn(),
        get: jest.fn(),
        has: jest.fn(),
        lock: jest.fn(),
        set: jest.fn(),
        isValidId: jest.fn(id => id === "valid"),
        generateId: jest.fn()
      };

      const setCookie = jest.fn();
      const clearCookie = jest.fn();

      let enhancedReq = enhanceRequestWithSession(
        {},
        cookieName,
        "invalid",
        sessions,
        setCookie,
        clearCookie
      );

      await enhancedReq.session.delete();
      expect(sessions.delete).not.toHaveBeenCalled();
      expect(setCookie).not.toHaveBeenCalled();
      expect(clearCookie).not.toHaveBeenCalled();
    });

    it("should work correctly if the provided cookie is valid", async () => {
      const sessions: SessionManager = {
        delete: jest.fn(),
        get: jest.fn(),
        has: jest.fn(),
        lock: jest.fn(),
        set: jest.fn(),
        isValidId: jest.fn(id => id === "valid"),
        generateId: jest.fn()
      };

      const setCookie = jest.fn();
      const clearCookie = jest.fn();

      let enhancedReq = enhanceRequestWithSession(
        {},
        cookieName,
        "valid",
        sessions,
        setCookie,
        clearCookie
      );

      await enhancedReq.session.delete();
      expect(sessions.delete).toHaveBeenCalledWith("valid");
      expect(setCookie).not.toHaveBeenCalled();
      expect(clearCookie).toHaveBeenCalledWith(cookieName);
    });
  });
});
