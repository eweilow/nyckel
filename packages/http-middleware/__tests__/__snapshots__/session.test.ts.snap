// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enhanceRequestWithSession should enhance a request correctly 1`] = `
Object {
  "delete": [Function],
  "get": [Function],
  "set": [Function],
}
`;
